service/storage_service.cc

```diff
@@ -880,9 +894,11 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
                     co_return;
                 }
                 co_await utils::get_local_injector().inject("ip-change-raft-sync-delay", std::chrono::milliseconds(500));
-                co_await _ss.mutate_token_metadata([this, hid](mutable_token_metadata_ptr t) {
-                    return _ss.sync_raft_topology_nodes(std::move(t), hid, {});
+                storage_service::nodes_to_notify_after_sync nodes_to_notify;
+                co_await _ss.mutate_token_metadata([this, hid, &nodes_to_notify](mutable_token_metadata_ptr t) -> future<> {
+                    nodes_to_notify = co_await _ss.sync_raft_topology_nodes(std::move(t), hid, {});
                 });
```

Comment:
This could cause notifications to get reordered, no?

Previously we would deliver them under token_metadata lock, so they would be delivered in the same order as corresponding token_metadata updates.

But now, if there's e.g. concurrent sync_raft_topology_nodes, one from topology_state_load, another from the gossiper subscriber, they both modify token metadata, but then `notify_nodes_after_sync` could be called in reverse order, I think

Reply:
the `notify_nodes_after_sync` call in topology_state_load is still delivered under the token_metadata lock, if my sight doesn't trick me

I guess the fact we do it outside the lock scope here is just an artifact of using `mutate_token_metadata`

We could:
- not use `mutate_token_metadata`
- or extend `mutate_token_metadata` with some post-replicate action (but still under the lock)

Reply:
>     * not use `mutate_token_metadata`

That's what I initially did, but dropped it because it was slightly more complicated and didn't think it mattered much - it doesn't for hints, it only needs to be triggered after the update.

I guess it would make sense if we had some listeners that don't preempt on notification and depended on the notification order.

Reply:
What about drivers? I vaguely recall we did have some issues that were due to drivers handling notifications delayed / in weird order...

Reply:
Hmm, right. Without it notifications about those events cannot work reliably. Doesn't mean they will (there might be other bugs), but it's a prerequisite. I'll change it so that notifications are triggered under the lock.

Reply:
Turns out that you can pass `acquire_merge_lock::no` to `mutate_token_metadata` and take the token metadata lock manually - which is what I did.